shared-bindings/busdevice/I2CDevice.c

```diff
@@ -0,0 +1,269 @@
+//|         with busio.I2C(SCL, SDA) as i2c:
+//|             device = I2CDevice(i2c, 0x70)
+//|             bytes_read = bytearray(4)
+//|             with device:
+//|                 device.readinto(bytes_read)
+//|             # A second transaction
+//|             with device:
+//|                 device.write(bytes_read)"""
+//|     ...
+//| 
+STATIC mp_obj_t busdevice_i2cdevice_make_new(const mp_obj_type_t *type, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    busdevice_i2cdevice_obj_t *self = m_new_obj(busdevice_i2cdevice_obj_t);
+    self->base.type = &busdevice_i2cdevice_type;
+    enum { ARG_i2c, ARG_device_address, ARG_probe };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_i2c, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_device_address, MP_ARG_REQUIRED | MP_ARG_INT },
+        { MP_QSTR_probe, MP_ARG_BOOL, {.u_bool = true} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    busio_i2c_obj_t* i2c = args[ARG_i2c].u_obj;
+
+    common_hal_busdevice_i2cdevice_construct(self, i2c, args[ARG_device_address].u_int, args[ARG_probe].u_bool);
+    if (args[ARG_probe].u_bool == true) {
+        common_hal_busdevice_i2cdevice___probe_for_device(self);
+    }
+
+    return (mp_obj_t)self;
+}
+
+STATIC mp_obj_t busdevice_i2cdevice_obj___enter__(mp_obj_t self_in) {
+    common_hal_busdevice_i2cdevice_lock(self_in);
+    return self_in;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(busdevice_i2cdevice___enter___obj, busdevice_i2cdevice_obj___enter__);
+
+STATIC mp_obj_t busdevice_i2cdevice_obj___exit__(size_t n_args, const mp_obj_t *args) {
+    common_hal_busdevice_i2cdevice_unlock(args[0]);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(busdevice_i2cdevice___exit___obj, 4, 4, busdevice_i2cdevice_obj___exit__);
+
+//| def readinto(self, buf, *, start=0, end=None):
+//|         """
+//|         Read into ``buf`` from the device. The number of bytes read will be the
+//|         length of ``buf``.
+//|         If ``start`` or ``end`` is provided, then the buffer will be sliced
+//|         as if ``buf[start:end]``. This will not cause an allocation like
+//|         ``buf[start:end]`` will so it saves memory.
+//|         :param bytearray buffer: buffer to write into
+//|         :param int start: Index to start writing at
+//|         :param int end: Index to write up to but not include; if None, use ``len(buf)``"""
+//|         ...
+//| 
+STATIC void readinto(busdevice_i2cdevice_obj_t *self, mp_obj_t buffer, int32_t start, mp_int_t end) {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(buffer, &bufinfo, MP_BUFFER_WRITE);
+
+    size_t length = bufinfo.len;
+    normalize_buffer_bounds(&start, end, &length);
+    if (length == 0) {
+        mp_raise_ValueError(translate("Buffer must be at least length 1"));
+    }
```

Comment:
This code fragment is repeated at least one other place and could be refactored into a single static checking routine, to save space. When you add SPI, I think there will be more places.

Reply:
The fragment only shows up in two places overall. I did try to refactor it into a routine and it saves 4 bytes. Not sure if it is worth it or not? But I did test it out so can easily commit it.

Reply:
@dhalbert Just wanting to check with you if I should change/leave this for the 4 bytes?

Reply:
I think it's fine as is. 4 bytes isn't worth it.